### DISCUSSION TITLE
`DataSourceGeneratorAttribute` display name formatting for classes

### DISCUSSION THREAD

**Discussion by VMelnalksnis** [thomhurst/TUnit#949] [Ideas]
As far as I can tell, `TUnit` just uses the argument type name in class names when a `DataSourceGeneratorAttribute` is used on a class. That means that I have to create a subclass for each fixture in order for it to have a different name in the test explorer, and it limits how the name can be formatted since it has to be a valid class name. In `NUnit` it is possible to change the display name with `TestFixtureData.SetArgDisplayNames`, maybe something similar is possible in `TUnit`?

**Comment by thomhurst** [comment]
Can you try in `0.1.1079`.

In your custom `DataSourceGeneratorAttribute`, implement the interface `ITestDiscoveryEvent`. Then on the `DiscoveredTestContext` object given to you, you can call `SetDisplayName`. You can then set it based on whatever logic you like.

For example:

```csharp
    [Test]
    [MyGenerator]
    public async Task PasswordTest(string password)
    {
        await Assert.That(TestContext.Current!.TestDetails.DisplayName).IsEqualTo("PasswordTest(REDACTED)");
    }
    
    public class MyGenerator : DataSourceGeneratorAttribute<string>, ITestDiscoveryEvent
    {
        public override IEnumerable<string> GenerateDataSources(DataGeneratorMetadata dataGeneratorMetadata)
        {
            yield return "Super Secret Password";
        }

        public void OnTestDiscovery(DiscoveredTestContext discoveredTestContext)
        {
            discoveredTestContext.SetDisplayName($"{discoveredTestContext.TestDetails.TestName}(REDACTED)");
        }
    }
```

**Reply by VMelnalksnis** [reply]
I'm not sure that's quite it - this changes the name of each test case instead of the class.
Here's an example of what I have in `NUnit` now:
![image](https://github.com/user-attachments/assets/92acda9f-be47-4a78-8adf-4af7ce7d00a1)
Where `OpenApiTests` is the class and `12.20-bookworm` is the display name of the fixture passed to the class. Using your what you suggested changes the name of each test case, `ApiDefinition_ShouldBeExpected(/swagger/v1/swagger.json)` in this case. It also overwrites the test case parameter values. What I'd like to format is the `12.20-bookworm` value in this example, instead of it just using the type name.

**Reply by thomhurst** [reply]
Got ya. Try this then:

Create a class inheriting from `ArgumentDisplayFormatter`. Here you define whether you support formatting specific objects, and then also the converting to a string value.

e.g. 
```csharp
public class MyFormatter : ArgumentDisplayFormatter
{
    public override bool CanHandle(object? value)
    {
        return value is SomeFixture;
    }

    public override string FormatValue(object? value)
    {
        var fixture = (SomeFixture) value;
        return fixture.SomeProperty;
    }
```

Then apply that to your test/class/assembly with the `ArgumentDisplayFormatter<>` attribute.

E.g.

```csharp
[ArgumentDisplayFormatter<MyFormatter>]
public class MyTests { ... }
```

And that should do it.

**Reply by VMelnalksnis** [reply]
Thanks, that works! Some related feedback - initially no tests were discovered, I'm guessing because my formatter was throwing an exception. However, there was no visible error in build output/IDE. Would it be possible to somehow surface exceptions such as this so that there is no need to guess what went wrong?
